File: mt76_session.py
# SPDX-License-Identifier: BSD-3-Clause-Clear
"""Bounded, single-owner acquisition after explicit firmware bring-up.

The worker owns USB; consumers own decoding/output. No warm adoption, automatic
recovery, transmission, or physical gap-free retuning is implied by this API.
"""

from __future__ import annotations

import math
import queue
import struct
import threading
import time
from collections import Counter, deque
from concurrent.futures import Future, TimeoutError
from dataclasses import dataclass

import usb.core

import mt7921u as m


class SessionError(RuntimeError):
    """Session stopped or faulted; a fresh bring-up is needed before restarting."""


@dataclass(frozen=True)
class Packet:
    raw: bytes
    received_ns: int
    epoch_ns: int
    channel_generation: int
    transitioning: bool
    kind: str


def packet_kind(raw: bytes, chip: str) -> str:
    """Classify bounded DMA records using the existing connac2/3 RX layouts.

    mt7921_queue_rx_skb / mt7925_queue_rx_skb at the pinned mt76 baseline;
    this is routing, not a replacement for full RX or MCU payload validation.
    """
    if len(raw) < 4:
        return "malformed"
    word = struct.unpack_from("<I", raw)[0]
    size = word & 0xFFFF
    if size < 4 or size > len(raw):
        return "malformed"
    kind = (word >> 27) & 31
    flag = (word >> m.RXD0_PKT_FLAG_SHIFT) & m.RXD0_PKT_FLAG_MASK
    if kind == m.PKT_TYPE_NORMAL or (kind == m.PKT_TYPE_RX_EVENT and flag == m.PKT_FLAG_NORMAL_MCU):
        return "frame" if size >= 32 else "malformed"
    if chip == "mt7925" and ((word >> 16) & 0x380F) == 0x3801:
        return "frame" if size >= 32 else "malformed"
    if kind == m.PKT_TYPE_RX_EVENT:
        return "reply" if size >= (44 if chip == "mt7925" else 36) else "malformed"
    return "status"


class AcquisitionSession:
    """One worker, bounded drop-newest queues, and serialized command callbacks.

    Start only after bringup/monitor/sniffer/tune on a caller-owned device. Then
    use call(lambda dev: ...) for short driver operations and read() for packets.
    Callbacks must not sleep, perform output, reenter the session, or reset/close
    the device. Python cannot forcibly cancel arbitrary user callback code.
    Stop the session before closing its device. Each new session requires bringup.
    """

    def __init__(self, dev, *, frame_capacity=256, event_capacity=64, command_capacity=16):
        for value in (frame_capacity, event_capacity, command_capacity):
            if not isinstance(value, int) or not 1 <= value <= 65536:
                raise ValueError("queue capacities must be integers in 1..65536")
        self.dev = dev
        self.frames = deque()
        self.events = deque()
        self.frame_capacity = frame_capacity
        self.event_capacity = event_capacity
        self.commands = queue.Queue(command_capacity)
        self.condition = threading.Condition()
        self.stop_event = threading.Event()
        self.counts = Counter()
        self.state = "new"
        self.error = None
        self.epoch_ns = 0
        self.generation = 0
        self.requested_channel = None
        self.transitioning = False
        self.deadline = None
        self.worker = None
        self.read_size = 16384
        self.poll_ms = 50

    def start(self):
        if self.state != "new" or getattr(self.dev, "_session", None) is not None:
            raise SessionError("device/session already owned")
        if not getattr(self.dev, "_session_ready", False) or not self.dev.evt_ep4:
            raise SessionError("explicit fresh bringup required before session start")
        self.epoch_ns = time.monotonic_ns()
        self.requested_channel = getattr(self.dev, "_capture_channel", None)
        self.dev._session = self
        self.dev._session_ready = False
        self.state = "running"
        self.worker = threading.Thread(target=self._run, name="mt76-acquisition", daemon=True)
        try:
            self.worker.start()
        except BaseException:
            self.dev._session = None
            self.state = "failed"
            raise
        return self

    def __enter__(self):
        return self.start()

    def __exit__(self, *_):
        self.stop()

    def check_owner(self):
        if threading.current_thread() is not self.worker:
            raise SessionError("USB is owned by the acquisition worker; use session.call/read")
        if self.stop_event.is_set():
            raise SessionError("session stopping")

    def io_timeout(self, requested_ms):
        self.check_owner()
        if self.deadline is None:
            return min(requested_ms, self.poll_ms)
        left = self.deadline - time.monotonic()
        if left <= 0:
            self._fail("command deadline expired; fresh bringup required")
            raise SessionError("command deadline expired; resynchronization required")
        return max(1, min(requested_ms, math.ceil(left * 1000)))

    def call(self, operation, *, timeout=3.0, retune=False):
        """Execute one short driver callback; timeout includes queueing and all USB I/O.

        Set retune=True for operations changing channel. Failure invalidates the
        session, not just the callback. Operation results may contain identifiers.
        """
        if not math.isfinite(timeout) or timeout <= 0:
            raise ValueError("timeout must be positive and finite")
        if threading.current_thread() is self.worker:
            raise SessionError("session.call cannot be reentered from a callback")
        future = Future()
        with self.condition:
            if self.state != "running":
                raise SessionError(self.error or "session is not running")
            self.commands.put_nowait((future, operation, time.monotonic() + timeout, retune))
        try:
            return future.result(timeout=timeout)
        except TimeoutError as exc:
            self._fail("command deadline expired; fresh bringup required")
            raise SessionError(self.error) from exc

    def tune(self, band, control, center=None, width=20, *, timeout=3.0):
        return self.call(
            lambda dev: dev.tune(band, control, center, width), timeout=timeout, retune=True
        )

    def read(self, *, timeout=1.0, events=False):
        """Return a Packet, or None on quiet timeout/after queued data is drained.

        Queued packets remain readable after stop/failure. snapshot() distinguishes
        end of session from a quiet channel. Raw frames are sensitive, never logged.
        """
        if threading.current_thread() is self.worker:
            raise SessionError("worker callbacks cannot consume their own session queues")
        if not math.isfinite(timeout) or timeout < 0:
            raise ValueError("timeout must be nonnegative and finite")
        end = time.monotonic() + timeout
        with self.condition:
            packets = self.events if events else self.frames
            while not packets:
                left = end - time.monotonic()
                if left <= 0 or self.state in ("closed", "failed", "new"):
                    return None
                self.condition.wait(left)
            packet = packets.popleft()
            self.counts["events_delivered" if events else "frames_delivered"] += 1
            return packet

    def snapshot(self):
        with self.condition:
            return {
                "state": self.state,
                "error": self.error,
                "epoch_ns": self.epoch_ns,
                "channel_generation": self.generation,
                "requested_channel": self.requested_channel,
                "frame_depth": len(self.frames),
                "event_depth": len(self.events),
                "counts": dict(self.counts),
            }

    def stop(self, timeout=2.0):
        if not math.isfinite(timeout) or timeout < 0:
            raise ValueError("stop timeout must be nonnegative and finite")
        if threading.current_thread() is self.worker:
            raise SessionError("worker cannot join itself")
        with self.condition:
            if self.state == "new":
                self.state = "closed"
            elif self.state == "running":
                self.state = "stopping"
            self.stop_event.set()
            self.condition.notify_all()
        if self.worker is not None:
            self.worker.join(timeout)
            if self.worker.is_alive():
                raise SessionError("worker still owns USB; do not close device")
        if getattr(self.dev, "_session", None) is self:
            self.dev._session = None

    def _fail(self, message):
        with self.condition:
            self.error = message
            self.state = "failed"
            self.requested_channel = None
            self.stop_event.set()
            self.condition.notify_all()

    def _count(self, key):
        with self.condition:
            self.counts[key] += 1

    def _receive(self, timeout):
        try:
            raw = bytes(self.dev.bulk_in(self.dev.ep_in_pkt_rx, self.read_size, timeout))
        except usb.core.USBTimeoutError:
            self._count("read_timeouts")
            return None
        except usb.core.USBError:
            self._count("usb_errors")
            self._fail("USB read failed; fresh bringup required")
            raise
        self._count("transfers")
        kind = packet_kind(raw, self.dev.CHIP)
        if kind == "malformed":
            self._count("malformed")
            return None
        # USB padding is not part of the DMA record.
        raw = raw[: struct.unpack_from("<H", raw)[0]]
        return Packet(
            raw, time.monotonic_ns(), self.epoch_ns, self.generation, self.transitioning, kind
        )

    def _enqueue(self, packet):
        frame = packet.kind == "frame"
        prefix = "frames" if frame else "events"
        with self.condition:
            self.counts[prefix + "_received"] += 1
            packets = self.frames if frame else self.events
            capacity = self.frame_capacity if frame else self.event_capacity
            if len(packets) == capacity:
                self.counts[prefix + "_dropped"] += 1
            else:
                packets.append(packet)
                self.counts[prefix + "_high_water"] = max(
                    self.counts[prefix + "_high_water"], len(packets)
                )
                self.condition.notify_all()

    def wait_reply(self, seq, cid, timeout):
        """Called only by the existing MCU encoder on the owner thread.

        Payload-specific validation remains in the command helper; event IDs do
        not generally equal command IDs. Never queue replies for future matching.
        A timeout faults the session before another sequence can be allocated.
        """
        self.check_owner()
        end = min(self.deadline, time.monotonic() + timeout / 1000)
        while time.monotonic() < end:
            self.check_owner()
            packet = self._receive(
                max(1, min(self.poll_ms, math.ceil((end - time.monotonic()) * 1000)))
            )
            if packet is None:
                continue
            if packet.kind == "reply":
                if packet.raw[self.dev.RXD_SEQ_OFFSET] == seq:
                    self._count("replies_matched")
                    return packet.raw
                self._count("unmatched_replies")
            self._enqueue(packet)
        self._fail("MCU reply deadline expired; fresh bringup required")
        raise SessionError(f"MCU command 0x{cid:02x} timed out; fresh bringup required")

    def _run(self):
        active = None
        try:
            while not self.stop_event.is_set():
                try:
                    active, operation, self.deadline, retune = self.commands.get_nowait()
                except queue.Empty:
                    packet = self._receive(self.poll_ms)
                    if packet:
                        if packet.kind == "reply":
                            self._count("unmatched_replies")
                        self._enqueue(packet)
                    continue
                self.check_owner()
                self.io_timeout(self.poll_ms)
                with self.condition:
                    self.transitioning = retune
                    if retune:
                        self.requested_channel = None
                result = operation(self.dev)
                self.io_timeout(self.poll_ms)
                with self.condition:
                    if retune:
                        self.generation += 1
                        self.requested_channel = getattr(self.dev, "_capture_channel", None)
                    self.transitioning = False
                self.deadline = None
                self._count("commands_completed")
                active.set_result(result)
                active = None
        except BaseException as exc:
            # Do not serialize exception text: user callbacks/USB errors may carry
            # identifiers. The original exception is returned only to the caller.
            if not self.stop_event.is_set():
                self._fail(f"{type(exc).__name__}: session failed; fresh bringup required")
            if active is not None:
                active.set_exception(exc)
        finally:
            with self.condition:
                if self.state != "failed":
                    self.state = "closed"
                while not self.commands.empty():
                    future, *_ = self.commands.get_nowait()
                    future.set_exception(SessionError("session ended before command ran"))
                self.condition.notify_all()

File: c/mt76_session.c
/* SPDX-License-Identifier: BSD-3-Clause-Clear */
#include "mt76_session.h"
#include <pthread.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

struct mt76_session {
    mt7921_dev_t *dev;
    pthread_t worker, owner;
    pthread_mutex_t lock;
    pthread_cond_t changed;
    bool ready, exited, joined, stopping;
    bool command_busy, command_done, retune, transitioning;
    int command_result;
    int (*operation)(mt7921_dev_t *, void *);
    void *context;
    uint64_t deadline_ns;
    unsigned frame_capacity, event_capacity, frame_head, event_head;
    mt_session_packet_t *frames, *events;
    mt_session_stats_t stats;
};

static uint64_t now_ns(void) {
    struct timespec t;
    clock_gettime(CLOCK_MONOTONIC, &t);
    return (uint64_t)t.tv_sec * 1000000000 + (uint64_t)t.tv_nsec;
}
static uint32_t le32(const uint8_t *p) {
    return (uint32_t)p[0] | (uint32_t)p[1] << 8 | (uint32_t)p[2] << 16 | (uint32_t)p[3] << 24;
}
static void wait_slice(mt76_session_t *s, uint64_t end) {
    uint64_t now = now_ns();
    if (now >= end) return;
    uint64_t ns = end - now;
    if (ns > 50000000) ns = 50000000;
    struct timespec delay = {(time_t)(ns / 1000000000), (long)(ns % 1000000000)};
    /* macOS relative wait avoids wall-clock adjustments changing deadlines. */
    pthread_cond_timedwait_relative_np(&s->changed, &s->lock, &delay);
}
static void fail_locked(mt76_session_t *s) {
    s->stopping = true;
    s->stats.state = MT_SESSION_FAILED;
    s->stats.channel_known = false;
    pthread_cond_broadcast(&s->changed);
}
static void fail(void *context) {
    mt76_session_t *s = context;
    pthread_mutex_lock(&s->lock);
    fail_locked(s);
    pthread_mutex_unlock(&s->lock);
}
static uint32_t io_timeout(void *context, uint32_t requested) {
    mt76_session_t *s = context;
    pthread_mutex_lock(&s->lock);
    uint32_t result = requested;
    if (!requested || !s->ready || !pthread_equal(pthread_self(), s->owner) || s->stopping) {
        result = 0;
    } else if (s->deadline_ns) {
        uint64_t now = now_ns();
        if (now >= s->deadline_ns) {
            fail_locked(s);
            result = 0;
        } else {
            uint64_t left = (s->deadline_ns - now + 999999) / 1000000;
            if (left < result) result = (uint32_t)left;
        }
    } else if (result > 50) result = 50;
    pthread_mutex_unlock(&s->lock);
    return result;
}

int mt_session_packet_kind(const uint8_t *raw, uint32_t len, int chip) {
    if (!raw || len < 4) return MT_PACKET_MALFORMED;
    uint32_t word = le32(raw), size = word & 0xFFFF;
    if (size < 4 || size > len) return MT_PACKET_MALFORMED;
    unsigned type = word >> 27, flag = (word >> RXD0_PKT_FLAG_SHIFT) & RXD0_PKT_FLAG_MASK;
    if (type == PKT_TYPE_NORMAL || (type == PKT_TYPE_RX_EVENT && flag == PKT_FLAG_NORMAL_MCU) ||
        (chip == MT_CHIP_MT7925 && ((word >> 16) & 0x380F) == 0x3801))
        return size >= 32 ? MT_PACKET_FRAME : MT_PACKET_MALFORMED;
    if (type == PKT_TYPE_RX_EVENT)
        return size >= (chip == MT_CHIP_MT7925 ? 44U : 36U) ? MT_PACKET_REPLY : MT_PACKET_MALFORMED;
    return MT_PACKET_STATUS;
}

static int receive(mt76_session_t *s, mt_session_packet_t *p, uint32_t timeout) {
    timeout = io_timeout(s, timeout);
    if (!timeout) return -1;
    p->len = sizeof(p->raw);
    int result = s->dev->mcu.read_bulk(&s->dev->usb, MT_ROLE_PKT_RX, p->raw, &p->len, timeout);
    pthread_mutex_lock(&s->lock);
    if (result == MT7921_ERR_TIMEOUT) {
        s->stats.read_timeouts++;
        pthread_mutex_unlock(&s->lock);
        return 1;
    }
    if (result) {
        if (!s->stopping) { s->stats.usb_errors++; fail_locked(s); }
        pthread_mutex_unlock(&s->lock);
        return -1;
    }
    s->stats.transfers++;
    p->kind = p->len > sizeof(p->raw) ? MT_PACKET_MALFORMED :
              mt_session_packet_kind(p->raw, p->len, s->dev->usb.chip);
    if (p->kind == MT_PACKET_MALFORMED) {
        s->stats.malformed++;
        pthread_mutex_unlock(&s->lock);
        return 1;
    }
    p->len = le32(p->raw) & 0xFFFF;
    p->received_ns = now_ns();
    p->epoch_ns = s->stats.epoch_ns;
    p->generation = s->stats.generation;
    p->transitioning = s->transitioning;
    pthread_mutex_unlock(&s->lock);
    return 0;
}
static void enqueue(mt76_session_t *s, const mt_session_packet_t *p) {
    pthread_mutex_lock(&s->lock);
    bool frame = p->kind == MT_PACKET_FRAME;
    unsigned capacity = frame ? s->frame_capacity : s->event_capacity;
    unsigned head = frame ? s->frame_head : s->event_head;
    uint32_t *depth = frame ? &s->stats.frame_depth : &s->stats.event_depth;
    uint64_t *high = frame ? &s->stats.frames_high_water : &s->stats.events_high_water;
    if (frame) s->stats.frames_received++; else s->stats.events_received++;
    if (*depth == capacity) {
        if (frame) s->stats.frames_dropped++; else s->stats.events_dropped++;
    } else {
        mt_session_packet_t *packets = frame ? s->frames : s->events;
        /* Copy only live bytes; never export uninitialized tail bytes. */
        mt_session_packet_t *target = &packets[(head + *depth) % capacity];
        target->len = p->len; target->received_ns = p->received_ns; target->epoch_ns = p->epoch_ns;
        target->generation = p->generation; target->transitioning = p->transitioning;
        target->kind = p->kind;
        memcpy(target->raw, p->raw, p->len);
        (*depth)++;
        if (*depth > *high) *high = *depth;
        pthread_cond_broadcast(&s->changed);
    }
    pthread_mutex_unlock(&s->lock);
}
static int wait_reply(void *context, uint8_t seq, uint8_t cid, uint8_t *reply,
                       uint32_t *reply_len, uint32_t timeout) {
    (void)cid; /* payload validation belongs to the command, EID need not equal CID */
    mt76_session_t *s = context;
    uint32_t budget = io_timeout(s, timeout);
    if (!budget) return -1;
    uint64_t end = now_ns() + (uint64_t)budget * 1000000;
    mt_session_packet_t packet;
    while (now_ns() < end) {
        uint64_t now = now_ns();
        if (now >= end) break;
        uint32_t left = (uint32_t)((end - now + 999999) / 1000000);
        int result = receive(s, &packet, left < 50 ? left : 50);
        if (result < 0) return -1;
        if (result) continue;
        if (packet.kind == MT_PACKET_REPLY) {
            if (packet.raw[s->dev->mcu.prof->rxd_seq_offset] == seq) {
                pthread_mutex_lock(&s->lock);
                s->stats.replies_matched++;
                pthread_mutex_unlock(&s->lock);
                if (reply && (!reply_len || *reply_len < packet.len)) { fail(s); return -1; }
                if (reply) { memcpy(reply, packet.raw, packet.len); *reply_len = packet.len; }
                return 0;
            }
            pthread_mutex_lock(&s->lock);
            s->stats.unmatched_replies++;
            pthread_mutex_unlock(&s->lock);
        }
        enqueue(s, &packet);
    }
    fail(s); /* even if a command helper swallows the error, no next sequence */
    return -1;
}
static void *run(void *context) {
    mt76_session_t *s = context;
    mt_session_packet_t packet;
    pthread_mutex_lock(&s->lock);
    s->owner = pthread_self(); s->ready = true;
    pthread_cond_broadcast(&s->changed);
    while (!s->stopping) {
        if (s->command_busy && !s->command_done) {
            s->transitioning = s->retune;
            if (s->retune) s->stats.channel_known = false;
            int (*operation)(mt7921_dev_t *, void *) = s->operation;
            void *arg = s->context;
            pthread_mutex_unlock(&s->lock);
            int result = io_timeout(s, 50) ? operation(s->dev, arg) : -1;
            if (!io_timeout(s, 50)) result = -1;
            pthread_mutex_lock(&s->lock);
            if (result) fail_locked(s);
            else {
                s->stats.commands_completed++;
                if (s->retune) {
                    s->stats.generation++;
                    s->stats.channel_known = s->dev->tuned;
                    s->stats.band = s->dev->tuned_band;
                    s->stats.control = s->dev->tuned_control;
                    s->stats.center = s->dev->tuned_center;
                    s->stats.width_mhz = s->dev->tuned_width;
                }
            }
            s->transitioning = false;
            s->deadline_ns = 0;
            s->command_result = result;
            s->command_done = true;
            pthread_cond_broadcast(&s->changed);
        } else {
            pthread_mutex_unlock(&s->lock);
            int result = receive(s, &packet, 50);
            if (!result) {
                if (packet.kind == MT_PACKET_REPLY) {
                    pthread_mutex_lock(&s->lock);
                    s->stats.unmatched_replies++;
                    pthread_mutex_unlock(&s->lock);
                }
                enqueue(s, &packet);
            }
            pthread_mutex_lock(&s->lock);
            if (result < 0 && !s->stopping) fail_locked(s);
        }
    }
    if (s->command_busy && !s->command_done) { s->command_result = -1; s->command_done = true; }
    if (s->stats.state != MT_SESSION_FAILED) s->stats.state = MT_SESSION_CLOSED;
    s->exited = true;
    pthread_cond_broadcast(&s->changed);
    pthread_mutex_unlock(&s->lock);
    return NULL;
}

mt76_session_t *mt_session_start(mt7921_dev_t *dev, unsigned frames, unsigned events) {
    if (!dev || !dev->session_ready || !dev->mcu.evt_ep4 || dev->usb.session_context ||
        !frames || frames > 4096 || !events || events > 4096) return NULL;
    mt76_session_t *s = calloc(1, sizeof(*s));
    if (!s) return NULL;
    s->frames = calloc(frames, sizeof(*s->frames));
    s->events = calloc(events, sizeof(*s->events));
    if (!s->frames || !s->events) goto bad;
    if (pthread_mutex_init(&s->lock, NULL)) goto bad;
    if (pthread_cond_init(&s->changed, NULL)) { pthread_mutex_destroy(&s->lock); goto bad; }
    s->dev = dev; s->frame_capacity = frames; s->event_capacity = events;
    s->stats.state = MT_SESSION_RUNNING; s->stats.epoch_ns = now_ns();
    s->stats.channel_known = dev->tuned;
    s->stats.band = dev->tuned_band; s->stats.control = dev->tuned_control;
    s->stats.center = dev->tuned_center; s->stats.width_mhz = dev->tuned_width;
    dev->session_ready = false;
    dev->usb.session_context = s; dev->usb.session_timeout = io_timeout; dev->usb.session_fail = fail;
    dev->mcu.session_context = s; dev->mcu.session_wait = wait_reply;
    if (pthread_create(&s->worker, NULL, run, s)) {
        dev->usb.session_context = NULL; dev->usb.session_timeout = NULL; dev->usb.session_fail = NULL;
        dev->mcu.session_context = NULL; dev->mcu.session_wait = NULL;
        pthread_cond_destroy(&s->changed); pthread_mutex_destroy(&s->lock);
        goto bad;
    }
    pthread_mutex_lock(&s->lock);
    while (!s->ready) pthread_cond_wait(&s->changed, &s->lock);
    pthread_mutex_unlock(&s->lock);
    return s;
bad:
    free(s->frames); free(s->events); free(s);
    return NULL;
}

int mt_session_call(mt76_session_t *s, int (*operation)(mt7921_dev_t *, void *),
                    void *context, uint32_t timeout_ms, bool retune) {
    if (!s || !operation || !timeout_ms) return -1;
    pthread_mutex_lock(&s->lock);
    if (s->stopping || pthread_equal(pthread_self(), s->owner)) {
        pthread_mutex_unlock(&s->lock); return -1;
    }
    if (s->command_busy) { pthread_mutex_unlock(&s->lock); return MT_SESSION_BUSY; }
    s->command_busy = true; s->command_done = false;
    s->operation = operation; s->context = context; s->retune = retune;
    s->deadline_ns = now_ns() + (uint64_t)timeout_ms * 1000000;
    while (!s->command_done) {
        if (now_ns() >= s->deadline_ns) {
            fail_locked(s);
            /* Must retain callback/context lifetime until the worker returns. */
            pthread_cond_wait(&s->changed, &s->lock);
        } else wait_slice(s, s->deadline_ns);
    }
    int result = s->command_result;
    s->command_busy = false;
    pthread_mutex_unlock(&s->lock);
    return result;
}

int mt_session_read(mt76_session_t *s, mt_session_packet_t *out, bool events, uint32_t timeout_ms) {
    if (!s || !out) return -1;
    uint64_t end = now_ns() + (uint64_t)timeout_ms * 1000000;
    pthread_mutex_lock(&s->lock);
    if (pthread_equal(pthread_self(), s->owner)) { pthread_mutex_unlock(&s->lock); return -1; }
    uint32_t *depth = events ? &s->stats.event_depth : &s->stats.frame_depth;
    while (!*depth && !s->stopping && now_ns() < end) wait_slice(s, end);
    if (!*depth) { pthread_mutex_unlock(&s->lock); return 1; }
    unsigned *head = events ? &s->event_head : &s->frame_head;
    unsigned capacity = events ? s->event_capacity : s->frame_capacity;
    mt_session_packet_t *packets = events ? s->events : s->frames;
    *out = packets[*head];
    *head = (*head + 1) % capacity; (*depth)--;
    if (events) s->stats.events_delivered++; else s->stats.frames_delivered++;
    pthread_mutex_unlock(&s->lock);
    return 0;
}
void mt_session_snapshot(mt76_session_t *s, mt_session_stats_t *out) {
    if (!s || !out) return;
    pthread_mutex_lock(&s->lock); *out = s->stats; pthread_mutex_unlock(&s->lock);
}
int mt_session_stop(mt76_session_t *s, uint32_t timeout_ms) {
    if (!s) return -1;
    uint64_t end = now_ns() + (uint64_t)timeout_ms * 1000000;
    pthread_mutex_lock(&s->lock);
    if (pthread_equal(pthread_self(), s->owner)) { pthread_mutex_unlock(&s->lock); return -1; }
    s->stopping = true;
    if (s->stats.state == MT_SESSION_RUNNING) s->stats.state = MT_SESSION_STOPPING;
    pthread_cond_broadcast(&s->changed);
    while (!s->exited && now_ns() < end) wait_slice(s, end);
    if (!s->exited) { pthread_mutex_unlock(&s->lock); return -1; }
    pthread_mutex_unlock(&s->lock);
    if (!s->joined) { pthread_join(s->worker, NULL); s->joined = true; }
    s->dev->usb.session_context = NULL; s->dev->usb.session_timeout = NULL;
    s->dev->usb.session_fail = NULL;
    s->dev->mcu.session_context = NULL; s->dev->mcu.session_wait = NULL;
    return 0;
}
int mt_session_destroy(mt76_session_t *s) {
    if (!s) return 0;
    pthread_mutex_lock(&s->lock);
    bool safe = s->joined && !s->command_busy;
    pthread_mutex_unlock(&s->lock);
    if (!safe) return -1;
    pthread_cond_destroy(&s->changed); pthread_mutex_destroy(&s->lock);
    free(s->frames); free(s->events); free(s);
    return 0;
}
